### ISSUE TITLE
Out of memory during MNIST model training.

### ISSUE THREAD

**Issue by bigdata2** [OpenMined/PySyft#1781]
I am training MNIST model with 2 virtual workers, my machine has 128 GB of memory, 97% of which is consumed by the python code (attached below). After 16th epoch, the process is killed by the OS. Has anyone faced similar issue before?

```
import copy
import torch
from torch import nn, optim
from torchvision import datasets, transforms
import torch.nn.functional as F
import argparse

# Training settings
parser = argparse.ArgumentParser(description='PyTorch MNIST Example')
parser.add_argument('--batch-size', type=int, default=64, metavar='N',
                    help='input batch size for training (default: 64)')
parser.add_argument('--test-batch-size', type=int, default=1000, metavar='N',
                    help='input batch size for testing (default: 1000)')
parser.add_argument('--epochs', type=int, default=10, metavar='N',
                    help='number of epochs to train (default: 10)')
parser.add_argument('--lr', type=float, default=0.01, metavar='LR',
                    help='learning rate (default: 0.01)')
parser.add_argument('--momentum', type=float, default=0.5, metavar='M',
                    help='SGD momentum (default: 0.5)')
parser.add_argument('--no-cuda', action='store_true', default=False,
                    help='disables CUDA training')
parser.add_argument('--seed', type=int, default=1, metavar='S',
                    help='random seed (default: 1)')
parser.add_argument('--log-interval', type=int, default=10, metavar='N',
                    help='how many batches to wait before logging training status')
args = parser.parse_args()
args.cuda = not args.no_cuda and torch.cuda.is_available()

class Net(nn.Module):
    def __init__(self):
        super(Net, self).__init__()
        self.conv1 = nn.Conv2d(1, 10, kernel_size=5)
        self.conv2 = nn.Conv2d(10, 20, kernel_size=5)
        self.conv2_drop = nn.Dropout2d()
        self.fc1 = nn.Linear(320, 50)
        self.fc2 = nn.Linear(50, 10)

    def forward(self, x):
        x = F.relu(F.max_pool2d(self.conv1(x), 2))
        x = F.relu(F.max_pool2d(self.conv2_drop(self.conv2(x)), 2))
        x = x.view(-1, 320)
        x = F.relu(self.fc1(x))
        x = F.dropout(x, training=self.training)
        x = self.fc2(x)
        return F.log_softmax(x, dim=1)

model = Net()
kwargs = {'num_workers': 1, 'pin_memory': True} if args.cuda else {}
train_loader = torch.utils.data.DataLoader(
    datasets.MNIST('../data', train=True, download=True,
                   transform=transforms.Compose([
                       transforms.ToTensor(),
                       transforms.Normalize((0.1307,), (0.3081,))
                   ])),
    batch_size=args.batch_size, shuffle=True, **kwargs)
test_loader = torch.utils.data.DataLoader(
    datasets.MNIST('../data', train=False, transform=transforms.Compose([
                       transforms.ToTensor(),
                       transforms.Normalize((0.1307,), (0.3081,))
                   ])),
    batch_size=args.test_batch_size, shuffle=True, **kwargs)

dataset_train = datasets.MNIST('../data/', train=True, download=True,
                       transform=transforms.Compose([
                       transforms.ToTensor(),
                       transforms.Normalize((0.1307,), (0.3081,))
                ]))


import syft as sy
hook = sy.TorchHook()
me = hook.local_worker
me.is_client_worker = False

# create a couple of socket workers and a virtual worker to aggregate parameters.
#alice = sy.SocketWorker(id="alice", hostname="52.71.29.35", port=2006, hook=hook, is_pointer=True, is_client_worker=False)
#bob   = sy.SocketWorker(id="bob", hostname="54.209.142.85", port=2005, hook=hook, is_pointer=True, is_client_worker=False)
bob = sy.VirtualWorker(id="bob")
alice = sy.VirtualWorker(id="alice")
secure_worker = sy.VirtualWorker(id="secure_worker")

bob.add_workers([alice, secure_worker])
alice.add_workers([bob, secure_worker])
secure_worker.add_workers([alice, bob])

compute_nodes = [bob, alice]
train_distributed_dataset  = []
for batch_idx, (data,target) in enumerate(train_loader):
    #if batch_idx > 4: break
    data = sy.Var(data)
    target = sy.Var(target.long())
    data.send(compute_nodes[batch_idx % len(compute_nodes)])
    target.send(compute_nodes[batch_idx % len(compute_nodes)])
    train_distributed_dataset.append((data, target))

bobs_model = model.copy().send(bob)
alices_model = model.copy().send(alice)

bobs_opt = optim.SGD(params=bobs_model.parameters(),lr=0.1)
alices_opt = optim.SGD(params=alices_model.parameters(),lr=0.1)

opt_nodes   = [bobs_opt, alices_opt]
model_nodes = [bobs_model, alices_model]

def train(epoch):
    #TODO train Bob and Alice in parallel.
    bobs_model = model.copy().send(bob)
    alices_model = model.copy().send(alice)
    for batch_idx, (data,target) in enumerate(train_distributed_dataset):

        bobs_opt = optim.SGD(params=bobs_model.parameters(),lr=0.1)
        alices_opt = optim.SGD(params=alices_model.parameters(),lr=0.1)

        opt_nodes[batch_idx % len(compute_nodes)].zero_grad()
        pred = model_nodes[batch_idx % len(compute_nodes)](data)
        loss = F.nll_loss(pred, target)
        loss.backward()

        opt_nodes[batch_idx % len(compute_nodes)].step()
        loss = loss.get().data[0]
        #print(loss)

    alices_model.move(secure_worker)
    bobs_model.move(secure_worker)

    model.conv1.weight.data.set_(((alices_model.conv1.weight.data + bobs_model.conv1.weight.data) / 2).get())
    model.conv1.bias.data.set_(((alices_model.conv1.bias.data + bobs_model.conv1.bias.data) / 2).get())
    model.conv2.weight.data.set_(((alices_model.conv2.weight.data + bobs_model.conv2.weight.data) / 2).get())
    model.conv2.bias.data.set_(((alices_model.conv2.bias.data + bobs_model.conv2.bias.data) / 2).get())
    model.fc1.weight.data.set_(((alices_model.fc1.weight.data + bobs_model.fc1.weight.data) / 2).get())
    model.fc1.bias.data.set_(((alices_model.fc1.bias.data + bobs_model.fc1.bias.data) / 2).get())
    model.fc2.weight.data.set_(((alices_model.fc2.weight.data + bobs_model.fc2.weight.data) / 2).get())
    model.fc2.bias.data.set_(((alices_model.fc2.bias.data + bobs_model.fc2.bias.data) / 2).get())

    len_train_loader = len(train_loader.dataset)
    if batch_idx % args.log_interval == 0:
       print('Train Epoch: {} [{}/{} ({:.0f}%)]\tLoss: {:.6f}'.format(
             epoch, batch_idx *  args.batch_size, len_train_loader * args.batch_size,
             100. * batch_idx / len_train_loader, loss))

def test():
    #model.eval()
    test_loss = 0
    correct = 0
    for data, target in test_loader:
        data, target = sy.Var(data), sy.Var(target)
        output = model(data)
        #print("output ", output)
        #print("loss ", F.nll_loss(output, target, size_average=False))
        test_loss += F.nll_loss(output, target, size_average=False).data[0] # sum up batch loss
        pred = output.data.max(1, keepdim=True)[1] # get the index of the max log-probability
        correct += pred.eq(target.data.view_as(pred)).long().cpu().sum()

    len_test_loader = len(test_loader.dataset)
    test_loss /= len_test_loader
    print('\nTest set: Average loss: {:.4f}, Accuracy: {}/{} ({:.0f}%)\n'.format(
        test_loss, correct, len_test_loader,
        100. * correct / len_test_loader))

for epoch in range(1, args.epochs + 1):
    print(epoch)
    train(epoch)
    test()
```

**Comment by LaRiffle** [comment]
Hi @bigdata2,
There are known issues with remote memory management (remote value is not destroyed when the pointer to it disappears), which are being addressed in the new PySyft version (compatible with PyTorch 1.0) which should be released in February.

**Comment by bigdata2** [comment]
Thanks for the update. I was also wondering if the version with PyTorch 1.0 will have a feature to let workers load training data separately? Specifically, the case where the entire training data is not read by the client first, but workers can read the training data independently from a local storage. Also, it would be nice to have a feature when workers can join a client asynchronously, i.e. a worker can indicate that it is now ready to join the federated learning pool of other workers and then take part in the model transmission and reception of averaged model parameters.

**Comment by iamtrask** [comment]
In the new master branch based on torch 1.0 - we have automatic garbage collection - which should mitigate this.

**Comment by bigdata2** [comment]
Hi, torch 1.0 does not work for me, I get the following exception;

> Process Process-1:
> Traceback (most recent call last):
>   File "/usr/local/lib/python3.6/dist-packages/syft/frameworks/torch/hook_args.py", line 104, in hook_function_args
>     hook_args = hook_method_args_functions[attr]
> KeyError: 'torch._C.set_num_threads'
> 
> During handling of the above exception, another exception occurred:
> 
> Traceback (most recent call last):
>   File "/usr/lib/python3.6/multiprocessing/process.py", line 258, in _bootstrap
>     self.run()
>   File "/usr/lib/python3.6/multiprocessing/process.py", line 93, in run
>     self._target(*self._args, **self._kwargs)
>   File "/usr/local/lib/python3.6/dist-packages/torch/utils/data/dataloader.py", line 111, in _worker_loop
>     torch.set_num_threads(1)
>   File "/usr/local/lib/python3.6/dist-packages/syft/frameworks/torch/hook.py", line 489, in overloaded_attr
>     response = TorchTensor.handle_func_command(command)
>   File "/usr/local/lib/python3.6/dist-packages/syft/frameworks/torch/tensors/interpreters/native.py", line 158, in handle_func_command
>     new_args, new_type = syft.frameworks.torch.hook_args.hook_function_args(cmd, args)
>   File "/usr/local/lib/python3.6/dist-packages/syft/frameworks/torch/hook_args.py", line 112, in hook_function_args
>     args, return_tuple=True
>   File "/usr/local/lib/python3.6/dist-packages/syft/frameworks/torch/hook_args.py", line 134, in build_hook_args_function
>     get_tensor_type_function = build_get_tensor_type(rule)
>   File "/usr/local/lib/python3.6/dist-packages/syft/frameworks/torch/hook_args.py", line 327, in build_get_tensor_type
>     return lambdas[0]
> IndexError: list index out of range
> Traceback (most recent call last):
>   File "/usr/local/lib/python3.6/dist-packages/syft/frameworks/torch/hook_args.py", line 104, in hook_function_args
>     hook_args = hook_method_args_functions[attr]
> KeyError: 'torch.randperm'
> 
> During handling of the above exception, another exception occurred:
> 
> Traceback (most recent call last):
>   File "test1.py", line 83, in <module>
>     for batch_idx, (data,target) in enumerate(train_loader):
>   File "/usr/local/lib/python3.6/dist-packages/torch/utils/data/dataloader.py", line 819, in __iter__
>     return _DataLoaderIter(self)
>   File "/usr/local/lib/python3.6/dist-packages/torch/utils/data/dataloader.py", line 584, in __init__
>     self._put_indices()
>   File "/usr/local/lib/python3.6/dist-packages/torch/utils/data/dataloader.py", line 646, in _put_indices
>     indices = next(self.sample_iter, None)
>   File "/usr/local/lib/python3.6/dist-packages/torch/utils/data/sampler.py", line 160, in __iter__
>     for idx in self.sampler:
>   File "/usr/local/lib/python3.6/dist-packages/torch/utils/data/sampler.py", line 73, in __iter__
>     return iter(torch.randperm(n).tolist())
>   File "/usr/local/lib/python3.6/dist-packages/syft/frameworks/torch/hook.py", line 489, in overloaded_attr
>     response = TorchTensor.handle_func_command(command)
>   File "/usr/local/lib/python3.6/dist-packages/syft/frameworks/torch/tensors/interpreters/native.py", line 158, in handle_func_command
>     new_args, new_type = syft.frameworks.torch.hook_args.hook_function_args(cmd, args)
>   File "/usr/local/lib/python3.6/dist-packages/syft/frameworks/torch/hook_args.py", line 112, in hook_function_args
>     args, return_tuple=True
>   File "/usr/local/lib/python3.6/dist-packages/syft/frameworks/torch/hook_args.py", line 134, in build_hook_args_function
>     get_tensor_type_function = build_get_tensor_type(rule)
>   File "/usr/local/lib/python3.6/dist-packages/syft/frameworks/torch/hook_args.py", line 327, in build_get_tensor_type
>     return lambdas[0]
> IndexError: list index out of range

**Comment by robert-wagner** [comment]
Hey @bigdata2 Could you send the code which throws that error?

**Comment by bigdata2** [comment]
Hi @robert-wagner here is the code.
```
import syft as sy
import copy
import torch
from torch import nn, optim
from torchvision import datasets, transforms
import torch.nn.functional as F
import argparse

# Training settings
parser = argparse.ArgumentParser(description='PyTorch MNIST Example')
parser.add_argument('--batch-size', type=int, default=64, metavar='N',
                    help='input batch size for training (default: 64)')
parser.add_argument('--test-batch-size', type=int, default=1000, metavar='N',
                    help='input batch size for testing (default: 1000)')
parser.add_argument('--epochs', type=int, default=10, metavar='N',
                    help='number of epochs to train (default: 10)')
parser.add_argument('--lr', type=float, default=0.01, metavar='LR',
                    help='learning rate (default: 0.01)')
parser.add_argument('--momentum', type=float, default=0.5, metavar='M',
                    help='SGD momentum (default: 0.5)')
parser.add_argument('--no-cuda', action='store_true', default=False,
                    help='disables CUDA training')
parser.add_argument('--seed', type=int, default=1, metavar='S',
                    help='random seed (default: 1)')
parser.add_argument('--log-interval', type=int, default=10, metavar='N',
                    help='how many batches to wait before logging training status')
args = parser.parse_args([])
args.cuda = not args.no_cuda and torch.cuda.is_available()

class Net(nn.Module):
    def __init__(self):
        super(Net, self).__init__()
        self.conv1 = nn.Conv2d(1, 10, kernel_size=5)
        self.conv2 = nn.Conv2d(10, 20, kernel_size=5)
        self.conv2_drop = nn.Dropout2d()
        self.fc1 = nn.Linear(320, 50)
        self.fc2 = nn.Linear(50, 10)

    def forward(self, x):
        x = F.relu(F.max_pool2d(self.conv1(x), 2))
        x = F.relu(F.max_pool2d(self.conv2_drop(self.conv2(x)), 2))
        x = x.view(-1, 320)
        x = F.relu(self.fc1(x))
        x = F.dropout(x, training=self.training)
        x = self.fc2(x)
        return F.log_softmax(x, dim=1)

model = Net()

kwargs = {'num_workers': 1, 'pin_memory': True} if args.cuda else {}
train_loader = torch.utils.data.DataLoader(
    datasets.MNIST('../data', train=True, download=True,
                   transform=transforms.Compose([
                       transforms.ToTensor(),
                       transforms.Normalize((0.1307,), (0.3081,))
                   ])),
    batch_size=args.batch_size, shuffle=True, **kwargs)
test_loader = torch.utils.data.DataLoader(
    datasets.MNIST('../data', train=False, transform=transforms.Compose([
                       transforms.ToTensor(),
                       transforms.Normalize((0.1307,), (0.3081,))
                   ])),
    batch_size=args.test_batch_size, shuffle=True, **kwargs)

dataset_train = datasets.MNIST('../data/', train=True, download=True,
                       transform=transforms.Compose([
                       transforms.ToTensor(),
                       transforms.Normalize((0.1307,), (0.3081,))
                ]))

hook = sy.TorchHook(torch)
# create a couple workers
bob = sy.VirtualWorker(hook, id="bob")
alice = sy.VirtualWorker(hook, id="alice")
secure_worker = sy.VirtualWorker(hook, id="secure_worker")

bob.add_workers([alice, secure_worker])
alice.add_workers([bob, secure_worker])
secure_worker.add_workers([alice, bob])

train_distributed_dataset  = []
for batch_idx, (data,target) in enumerate(train_loader):
    if batch_idx > 4: break
    data = sy.Var(data)
    target = sy.Var(target.long())
    data.send(bob)
    target.send(bob)
    train_distributed_dataset.append((data, target))

bobs_model = Net()
bobs_opt = optim.SGD(params=bobs_model.parameters(),lr=0.1)
bobs_model = bobs_model.send(bob)

for batch_idx, (data,target) in enumerate(train_distributed_dataset):

    print(data)
    bobs_model.send(data.location)
    # Train Bob's Model
    bobs_opt.zero_grad() 
    bobs_pred = bobs_model(data)
    bobs_loss = F.nll_loss(bobs_pred, target)
    bobs_loss.backward()
    bobs_opt.step()
    bobs_loss = bobs_loss.get().data[0]
```

**Comment by jlebensold** [comment]
@robert-wagner I think this is related to `randperm`. Is it possible we're not hooking it correctly?

**Comment by LaRiffle** [comment]
Hi @bigdata2 
This is now solved! We have made a notebook tutorial for this case: https://github.com/OpenMined/PySyft/blob/dev/examples/tutorials/Part%208%20-%20Federated%20Learning%20on%20MNIST%20using%20a%20CNN.ipynb

**Comment by bigdata2** [comment]
I am still getting the same error messages and exceptions as reported originally.
